### PR TITLE
add subdirectories to screenshots

### LINKED ISSUE #8711: Screenshot Subfolders
**Is your feature request related to a problem? Please describe.**
After playing for just over a month, I have over 2,300 screenshots in my screenshot folder. Windows is beginning to struggle to load the folder, and actually finding a given screenshot is an absolute nightmare.

**Describe the solution you'd like**
The screenshots should be organised into subfolders. 
The actual subfolders themselves are fairly easy to decide, simply separate each of the selectable categories in the plugin into their own respective folders.

- Manual Screenshots
- Rewards
- Levels
- Kingdom Reward
- Pet
- PvP Kills
- Boss Kills
- Deaths
- Duels
- Valuable Drops
- Untradable Drops

Levels could also potentially be divided into further subfolders for each skill, since theoretically a maxed character could have 2245 screenshots in their levels folder. 

This would not only make searching for a specific screenshot much easier, it would massively reduce the loading time of the folder and the indexing when viewing it on windows.

**Describe alternatives you've considered**
You could do this manually, but that's inconvenient

### PR DESCRIPTION
**Pull request by gregg1494**
Closes #8711
Created directories for:
Manual Screenshots
Clue Scroll Rewards
Levels
Kingdom Rewards
Pets
PvP Kills
Boss Kills
Deaths
Duels
Valuable Drops
Untradable Drops
Quests

I was able to take actual screenshots for some of the basic ones like Levels, Deaths, Duels, and Manual Screenshots as those were easiest to do in game, and they worked correctly. I used ScreenshotPluginTest and check the return values for my screenShotSubDir variable and they are correct. I was not able to have the test take actual screen shots though.  

If someone knows how to actually take a screen shot when using the test calls let me know.

### CODE REVIEW

File: runelite-client/src/test/java/net/runelite/client/plugins/screenshot/ScreenshotPluginTest.java [hunk 1]
```diff
@@ -113,8 +111,14 @@ public void testClueScroll()
...
 		screenshotPlugin.onChatMessage(chatMessageEvent);
 
-		assertEquals("medium", screenshotPlugin.getClueType());
-		assertEquals(28, screenshotPlugin.getClueNumber());
+		WidgetLoaded event = new WidgetLoaded();
+		event.setGroupId(CLUE_SCROLL_REWARD_GROUP_ID);
+		screenshotPlugin.onWidgetLoaded(event);
+
+
+		//assertEquals("medium", screenshotPlugin.getClueType());
```
Comment (Nightfirecat) [inline runelite-client/src/test/java/net/runelite/client/plugins/screenshot/ScreenshotPluginTest.java]: Why are you commenting out old tests? This is not the right way to change the test suite for this plugin.

Comment (gregg1494) [reply runelite-client/src/test/java/net/runelite/client/plugins/screenshot/ScreenshotPluginTest.java]: I commented it out to test if my implementation was correct. Didn't see the need to create a more tests when I could just check my value from there. I initially didn't comment it out but didn't know if having two assertEquals would mess up the check.

Comment (Nightfirecat) [reply runelite-client/src/test/java/net/runelite/client/plugins/screenshot/ScreenshotPluginTest.java]: You can have any number of JUnit assertions in a single test case. The only important thing is that none of them fail.

Comment (gregg1494) [reply runelite-client/src/test/java/net/runelite/client/plugins/screenshot/ScreenshotPluginTest.java]: Ok. Thanks for clarifying.

File: runelite-client/src/test/java/net/runelite/client/plugins/screenshot/ScreenshotPluginTest.java [hunk 2]
```diff
@@ -36,8 +36,8 @@
 import net.runelite.api.events.GameTick;
 import net.runelite.api.events.WidgetLoaded;
 import net.runelite.api.widgets.Widget;
-import static net.runelite.api.widgets.WidgetID.DIALOG_SPRITE_GROUP_ID;
-import static net.runelite.api.widgets.WidgetID.LEVEL_UP_GROUP_ID;
+
```
Comment (Nightfirecat) [inline runelite-client/src/test/java/net/runelite/client/plugins/screenshot/ScreenshotPluginTest.java]: Revert the changes to this file.

File: runelite-client/src/test/java/net/runelite/client/plugins/screenshot/ScreenshotPluginTest.java [hunk 3]
```diff
@@ -36,8 +36,7 @@
 import net.runelite.api.events.GameTick;
 import net.runelite.api.events.WidgetLoaded;
 import net.runelite.api.widgets.Widget;
-import static net.runelite.api.widgets.WidgetID.DIALOG_SPRITE_GROUP_ID;
-import static net.runelite.api.widgets.WidgetID.LEVEL_UP_GROUP_ID;
+import static net.runelite.api.widgets.WidgetID.*;
```
Comment (Nightfirecat) [inline runelite-client/src/test/java/net/runelite/client/plugins/screenshot/ScreenshotPluginTest.java]: These changes still aren't reverted. Don't use star imports here. The full diff should show no differences in this test file.

File: runelite-client/src/main/java/net/runelite/client/plugins/screenshot/ScreenshotPlugin.java [hunk 4]
```diff
@@ -751,4 +776,10 @@ int gettheatreOfBloodNumber()
 	{
 		return theatreOfBloodNumber;
 	}
+
+	@VisibleForTesting
+	String getScreenShotSubDir()
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenshot/ScreenshotPlugin.java]: I guess this can be removed since it's no longer tested?

File: runelite-client/src/main/java/net/runelite/client/plugins/screenshot/ScreenshotPlugin.java [hunk 5]
```diff
@@ -130,6 +130,8 @@
 		"You feel something weird sneaking into your backpack",
 		"You have a funny feeling like you would have been followed");
 
+	private String screenShotSubDir = "Manual Screenshots";
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenshot/ScreenshotPlugin.java]: Instead of maintaining a field for the screenshot subdirectory, add a new signature to `takeScreenshot` which accepts a filename, subdirectory, and image. That way, these extra `screenShotSubDir = "Manual Screenshots";` resets can be removed and screenshot function calls can be more declarative.

File: runelite-client/src/main/java/net/runelite/client/plugins/screenshot/ScreenshotPlugin.java (hunk 4 above)
Comment (gregg1494) [reply runelite-client/src/main/java/net/runelite/client/plugins/screenshot/ScreenshotPlugin.java]: Ya that is no longer need. I will remove it.

File: runelite-client/src/main/java/net/runelite/client/plugins/screenshot/ScreenshotPlugin.java (hunk 5 above)
Comment (gregg1494) [reply runelite-client/src/main/java/net/runelite/client/plugins/screenshot/ScreenshotPlugin.java]: I still need a variable to track the the sub directory, are you suggesting just make it local to the function in which the subdirectory is created? I couldn't just pass the subdirectory name (for reasons like in the code below) where the takeScreenshot is called after a switch statement so I would need store the subdirectory based on the switch case.

https://github.com/runelite/runelite/blob/150d8722a889df943a5e3f7bc9811e863dc60d87/runelite-client/src/main/java/net/runelite/client/plugins/screenshot/ScreenshotPlugin.java#L517-L555

Comment (Nightfirecat) [reply runelite-client/src/main/java/net/runelite/client/plugins/screenshot/ScreenshotPlugin.java]: I agree, a variable will be needed for the `GameTick` and `WidgetLoaded` subscribers, however this doesn't need to be a class-wide field.

Comment (gregg1494) [reply runelite-client/src/main/java/net/runelite/client/plugins/screenshot/ScreenshotPlugin.java]: Ok I will see what I can do.

File: runelite-client/src/main/java/net/runelite/client/plugins/screenshot/ScreenshotPlugin.java [hunk 6]
```diff
@@ -263,27 +263,31 @@ public void onGameTick(GameTick event)
 			return;
 		}
 
+		String screenshotSubDir = "";
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenshot/ScreenshotPlugin.java]: Set this to `null`, and null-check it along with fileName before calling `takeScreenshot()`.

File: runelite-client/src/main/java/net/runelite/client/plugins/screenshot/ScreenshotPlugin.java [hunk 7]
```diff
@@ -563,8 +573,9 @@ String parseLevelUpWidget(WidgetInfo levelUpLevel)
 	 * and optionally uploads it to an image-hosting service.
 	 *
 	 * @param fileName    Filename to use, without file extension.
+     * @param subDir    Sub directory where the file will be put.
```
[the remark this answers is not in the document]
Comment (gregg1494) [reply runelite-client/src/main/java/net/runelite/client/plugins/screenshot/ScreenshotPlugin.java]: I think I fixed it. I cant tell my check style is saying its good.

[the remark this answers is not in the document]
Comment (Nightfirecat) [reply runelite-client/src/main/java/net/runelite/client/plugins/screenshot/ScreenshotPlugin.java]: It's not fixed, it's still 5 spaces instead of `\t `.

File: runelite-client/src/main/java/net/runelite/client/plugins/screenshot/ScreenshotPlugin.java [hunk 8]
```diff
@@ -563,8 +573,9 @@ String parseLevelUpWidget(WidgetInfo levelUpLevel)
 	 * and optionally uploads it to an image-hosting service.
 	 *
 	 * @param fileName    Filename to use, without file extension.
+	 * @param subDir    Sub directory where the file will be put.
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenshot/ScreenshotPlugin.java]: To line up with the javadoc above it.
```suggestion
	 * @param subDir      Sub directory where the file will be put.
```

**Comment by renfc01** [comment]
Hi, is this still being worked on?
